cookbook/en/making-http-requests.md

```diff
@@ -0,0 +1,367 @@
+            }
+
+            $data = json_decode($response->getBody()->getContents(), true, 512, JSON_THROW_ON_ERROR);
+            return $data;
+        } catch (ClientExceptionInterface $e) {
+            $this->logger->error('HTTP client error when fetching weather data', [
+                'city' => $city,
+                'error' => $e->getMessage(),
+            ]);
+            return null;
+        } catch (\JsonException $e) {
+            $this->logger->error('Failed to decode weather API response', [
+                'city' => $city,
+                'error' => $e->getMessage(),
+            ]);
+            return null;
+        }
+    }
+}
+```
+
+## Advanced usage
+
+### Using middlewares
+
+Guzzle supports middleware for cross-cutting concerns like authentication, logging, or retrying:
+
+```php
+<?php
+
+declare(strict_types=1);
+
+use GuzzleHttp\Client;
+use GuzzleHttp\HandlerStack;
+use GuzzleHttp\Middleware;
+use Psr\Http\Message\RequestInterface;
+use Psr\Log\LoggerInterface;
+
+class HttpClientFactory
+{
+    public function __construct(
+        private LoggerInterface $logger,
+    ) {
+    }
+
+    public function createClient(): Client
+    {
+        $stack = HandlerStack::create();
+        
+        // Add request/response logging middleware
+        $stack->push(Middleware::log(
+            $this->logger,
+            new \GuzzleHttp\MessageFormatter('HTTP {method} {uri} - {code} {phrase}')
+        ));
+        
+        // Add retry middleware
+        $stack->push(Middleware::retry(
+            function (int $retries, RequestInterface $request) {
+                return $retries < 3;
+            }
+        ));
+
+        return new Client([
+            'handler' => $stack,
+            'timeout' => 30,
+        ]);
+    }
+}
+```
+
+### Async requests
+
+For better performance when making multiple requests, you can use asynchronous requests:
+
+```php
+<?php
+
+declare(strict_types=1);
+
+use GuzzleHttp\Client;
+use GuzzleHttp\Promise\PromiseInterface;
+use Psr\Http\Message\ResponseInterface;
+
+class BatchApiService
+{
+    public function __construct(
+        private Client $httpClient,
+    ) {
+    }
+
+    /**
+     * @param array<int> $userIds
+     * @return array<ResponseInterface>
+     */
+    public function fetchMultipleUsers(array $userIds): array
+    {
+        $promises = [];
+        
+        foreach ($userIds as $userId) {
+            $promises[$userId] = $this->httpClient->getAsync(
+                "https://jsonplaceholder.typicode.com/users/{$userId}"
+            );
+        }
+
+        // Wait for all requests to complete
+        $responses = \GuzzleHttp\Promise\settle($promises)->wait();
+        
+        $results = [];
+        foreach ($responses as $userId => $response) {
+            if ($response['state'] === PromiseInterface::FULFILLED) {
+                $results[$userId] = $response['value'];
+            }
+        }
+        
+        return $results;
+    }
+}
+```
```

Comment:
Add a note that this feature isn't part of PSR interfaces, so the code will depend on Guzzle explicitly.